bondjimbond/islandora_marcxml
# Update transform

**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-2411)

* Other Relevant Links (https://github.com/Islandora/islandora_marcxml/pull/64)

# What does this Pull Request do?

Enables users to substitute their own xslt files for the MODS to MARC and MARC to MODS transforms.

# What's new?

* When the select_or_other module is installed, the xslt selection widgets are changed to enable the user to enter provide file paths to any file under the drupal root to be substituted for use for the transform. The user sees a fourth option: "Other (path to file from site root)". When selected, a text field appears where the path can be entered. 
* A validate handler ensures that the file exists when saving the form.
* All places that utilize the saved xslt path variable call `_islandora_marcxml_xslt_file_path()` to get the actual path to the file. This function converts the built-in xslt selections to drupal-root relative paths, and assumes that user-provided paths are drupal-root relative.

# How should this be tested?

* Test for normal behavior of the admin form and transform behavior with this change applied but without the select_or_other module installed.
  * Do the configuration settings appear and save correctly?
  * Do the selected xslt transforms apply correctly?
* Install the select_or_other module
* Try saving various islandora_marcxml xslt file options and testing that they save appropriately and apply correctly during MODS to MARC and MARC to MODS transforms:
  * Standard xslt files provided by the module.
  * A path to a custom xslt file provided by the user.
  * An invalid path to a file (should prevent submission of the admin form).

# Additional Notes:
* Does this change the interface, add a new feature, or otherwise change behaviours that would require updating documentation? 
   * Yes, the documentation is updated.
* Does this change add any new dependencies? 
  * The select_or_other module is a dependency in order to take advantage of the custom xslts option.
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? 
  * No.
* Could this change impact execution of existing code?
   * Yes, the functions which apply the xslt transforms utilize a different method for determining the paths to the xslt files.

# Interested parties
Tag (@bondjimbond @jordandukart
